### PR TITLE
Game engine

### LINKED ISSUE #9: BE: Game Engine V1
- Responsible for determining the next state. A game engine is essentially a class: https://dzone.com/articles/how-to-create-instantiate-a-class-in-nodejs 
- Two teams: red and blue team 
- Game class that initializes a game given two players. It initializes a Board with random words from a dictionary. It also assigns roles to the players (spymaster and guesser)
  - The dictionary could be something that’s hard coded
- Game State (Board state, whose turn it is, points)
  - Board state
  - Whose turn it is (red or blue)
  - Points (red points and blue points)
- methods for next move : given a player and a move, figure out the next state and return
  -If a player makes a move (for example: guesses one word), determine the next state (for ex: dependant on a bunch of other rules)
- method for determining if match is over
  - If you click on a black card 
  - If you win the game (one team guesses all their words) 
- method for restarting
  - All the words get reshuffled again
  -Points reset to 0

### LINKED ISSUE #19: BE: Game Engine V2
- to_json method : include all the necessary information relayed to the frontend
- assign method: assign one user to one role so that it can be relayed to the frontend

### PR DESCRIPTION
**Pull request by mpmpx**
Game engine V1 supports:
- Create a game with fixed four players.
- Next-move method for spies.
- Next-move method for guessers.
- End-turn method
- Restart method

Utility methods and dictionary methods.

### CODE REVIEW

**Review by Travmatth — COMMENTED** [review]
+1

File: server/game/utils/shuffle.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  // force size to be size of array if it is larger than the size of array.
+  if (size > array.length) size = array.length;
+
+  // swap arr[i] and arr[j] in every iteration where j is random from 0 to i.
+  for (let i = array.length - 1; i >= array.length - size; i--) {
+    let j = Math.floor(Math.random() * (i + 1));
+    let tmp = array[i];
+    array[i] = array[j];
+    array[j] = tmp;
+  }
```
Comment (sina-jamshidi) [inline server/game/utils/shuffle.js:19]: cool!

File: server/game/game.js [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+  this.gameState = {
+    playerList: new Set(playerList),
+    waitingList: playerList,
+    redSpy: undefined,
+    redGuessers: [],
+    blueSpy: undefined,
+    blueGuessers: [],
+    isReady: false,
+    isStart: false,
```
Comment (sina-jamshidi) [inline server/game/game.js:52]: Good start

**Review by sina-jamshidi — APPROVED** [review]
Looks good to me!
This kind of thing is a little hard to make sure is right until we try it out in a game. So bugs will probably come up at some point but looks good for now!

**Pull request merged**